Add angular material (2+)

srijanreddy98 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following:

- Adds @angular/material for material usage in angular2+.

### How to use:
- Please visit https://material.angular.io/components/categories to check out all the components.
- When looking at a component please look at the api's tab
![image](https://user-images.githubusercontent.com/23002114/86085712-6e47b580-babd-11ea-900b-a269045c6bbd.png)
- Check if the module exists in file called material.module.ts (three places: has been imported, in imports array and in exports array.
![image](https://user-images.githubusercontent.com/23002114/86085837-cf6f8900-babd-11ea-8ebc-2521fa19dfa3.png)

- If not present, then just add it there.

**Note: The material css overrides oppia.css, please make your classes more specific when you use material directives.**




## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

srijanreddy98 committed:
Introduce angular material

srijanreddy98 committed:
Add angular material

srijanreddy98 committed:
CODEOWNERS

srijanreddy98 committed:
hammer js

srijanreddy98 committed:
import hammer.js for animations

srijanreddy98 committed:
chang import order

srijanreddy98 commented:
@vojtechjelinek, @ankita240796, @DubeySandeep, and @bansalnitish ptal!

ankita240796 reviewed (approved):
LGTM for App.ts, Thanks @srijanreddy98!

bansalnitish reviewed (approved):
Wow! thanks @srijanreddy98

bansalnitish commented:
`ci/circleci: e2e_non_prod_tests` looks flaky

DubeySandeep commented on core/templates/App.ts:
```diff
@@ -15,7 +15,8 @@
  * @fileoverview Initialization and basic configuration for the Oppia module.
  */
-
+require('@angular/material/prebuilt-themes/deeppurple-amber.css');
```
Do we had this theme earlier? Do we have any alternatives?

srijanreddy98 replied:
No, we didn't. The reason to import this theme was to get material css. There are four prebuilt themes but I suggest we don't go for them, instead, we should do this: https://material.angular.io/guide/theming-your-components#theming-your-custom-component-with-angular-materials-theming-system.

DubeySandeep replied:
But that need Sass which we don't use as of now, right? ANy other alternatives. Also, can you explain the impact of choosing a theme here (with some UI examples).

srijanreddy98 replied:
There is no immediate impact, however, if we use the material directives, this CSS may override the oppia.css. In that case, we should be make our CSS rules more specific.

srijanreddy98 replied:
> ANy other alternatives.

This is the best material design UI library for angular2+ (Made by the angular team for angular 2+). I would say first-party over third party libraries any day.

srijanreddy98 replied:
Also, please note that this change has been brought with the intention of easing migration of angularjs material components as this library is very similar to angularjs material.

DubeySandeep commented on core/templates/App.ts:
```diff
@@ -15,7 +15,8 @@
  * @fileoverview Initialization and basic configuration for the Oppia module.
  */
-
+require('@angular/material/prebuilt-themes/deeppurple-amber.css');
```
Add a newline above*

srijanreddy98 replied:
Done!

DubeySandeep commented on core/templates/components/material.module.ts:
```diff
@@ -0,0 +1,31 @@
+import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+
+@NgModule({
+  imports: [CommonModule, BrowserAnimationsModule,
```
Add this is in sorted order*

srijanreddy98 replied:
Done!

DubeySandeep commented on core/templates/components/material.module.ts:
```diff
@@ -0,0 +1,31 @@
+  imports: [CommonModule, BrowserAnimationsModule,
+    MatCardModule, MatButtonModule
+  ],
+  exports: [BrowserAnimationsModule, MatCardModule, MatButtonModule]
```
Ditto*

srijanreddy98 replied:
Done!

DubeySandeep commented on core/templates/components/shared-component.module.ts:
```diff
@@ -23,14 +23,15 @@ import { CommonModule } from '@angular/common';
   declarations: [BackgroundBannerComponent, TranslatePipe],
   entryComponents: [BackgroundBannerComponent],
-  exports: [BackgroundBannerComponent, TranslatePipe]
+  exports: [MaterialModule, BackgroundBannerComponent, TranslatePipe]
```
Ditto*

srijanreddy98 replied:
Done!

DubeySandeep reviewed (commented):
LGTM, left a few minor comments/questions PTAL!

DubeySandeep commented:
@srijanreddy98, Thanks for adding the "How to use section", do we want contributors to follow a guideline while using material module? IF yes then should we move this "How to use" section to a (shareable) wiki page?

vojtechjelinek commented on core/templates/App.ts:
```diff
@@ -15,7 +15,8 @@
  */
-
+require('@angular/material/prebuilt-themes/deeppurple-amber.css');
+import 'hammerjs';
```
Why is this added?

srijanreddy98 replied:
angular material (v8) requires hammerjs for gesture recognition. You can look at point 2 [here](https://v8.material.angular.io/guide/getting-started#install-angular-material). But angular material (v9) onwards doesn't require hammerjs. So mostly this is will be temporary.

vojtechjelinek replied:
Okay, please add a TODO here referencing #9172 and containing explanation why is this needed.

srijanreddy98 replied:
Done!

vojtechjelinek reviewed (changes requested):
LGTM! Just one question.

srijanreddy98 committed:
Changes

srijanreddy98 committed:
NEw line

srijanreddy98 commented:
> @srijanreddy98, Thanks for adding the "How to use section", do we want contributors to follow a guideline while using material module? IF yes then should we move this "How to use" section to a (shareable) wiki page?

@DubeySandeep yes, I will move it to the wiki. Also, I have addressed your comments, PTAL!

srijanreddy98 commented:
@vojtechjelinek I have addressed your comments too. PTAL!

srijanreddy98 committed:
Comments

srijanreddy98 commented:
@vojtechjelinek Addressed your comments, PTAL!

vojtechjelinek commented on core/templates/App.ts:
```diff
@@ -16,6 +16,15 @@
 
+require('@angular/material/prebuilt-themes/deeppurple-amber.css');
+/**
+ * Angular material (v8) requires hammerjs for gesture recognition.
+ * You can look at point 2 here:
+ * https://v8.material.angular.io/guide/getting-started#install-angular-material
+ * It won't be reequired in angular 9.
+ * TODO(#9172): Remove the import when upgraded to angular 9.
```
```suggestion
 * Angular Material (v8) requires hammerjs for gesture recognition.
 * You can look at point 2 here:
 * https://v8.material.angular.io/guide/getting-started#install-angular-material
 * It won't be required in Angular 9.
 * TODO(#9172): Remove the import when upgraded to Angular 9.
```

vojtechjelinek reviewed (approved):
LGTM! Just one nit.

srijanreddy98 committed:
Update core/templates/App.ts

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

srijanreddy98 committed:
code review

srijanreddy98 committed:
Merge branch 'angular-material' of https://github.com/srijanreddy98/oppia into angular-material

DubeySandeep reviewed (approved):
LGTM, thansk for the work, @srijanreddy98! :)

srijanreddy98 committed:
Changes

srijanreddy98 committed:
Changes
